File: crates/trailryx-projection/tests/oracle.py
"""Read a Trailryx projection with pyarrow and check every cell.

Deliberately not our code. The Parquet writer in this repository is
hand-written, and a hand-written encoder checked only by its author is an
encoder that is Parquet-shaped rather than Parquet. This script takes the file
and a table of what the writer believed it wrote, and makes somebody else's
reader settle the argument.

Usage: oracle.py <parquet file> <expected tsv>

The TSV is one line per cell: column, row, value, with a leading marker for a null
so a genuinely empty string stays distinguishable.

A list column's cell is the elements joined with commas, and the reader is required
to hand back an actual list for it rather than a string. That direction of the check
is the point: the columns used to be comma-joined strings, and a reader that returned
a string would now be reading a file whose schema says list. Comparing only the
rendered text would have passed either way.
"""

import sys
import pyarrow as pa
import pyarrow.parquet as pq

NULL = "\\0NULL"
LIST_MARKER = "\\0LIST"


def main() -> int:
    table = pq.read_table(sys.argv[1])
    expected = {}
    with open(sys.argv[2], encoding="utf-8") as handle:
        for line in handle:
            column, row, value = line.rstrip("\n").split("\t", 2)
            expected[(column, int(row))] = value

    columns = set(table.column_names)
    wanted = {c for c, _ in expected}
    if columns != wanted:
        print(f"columns differ: only in file {columns - wanted}, only expected {wanted - columns}")
        return 1

    checked = 0
    lists = 0
    for name in table.column_names:
        field = table.schema.field(name)
        values = table.column(name).to_pylist()
        is_list = pa.types.is_list(field.type) or pa.types.is_large_list(field.type)

        # A column the writer says is a list must arrive as one. This is the half of
        # the check that a text comparison cannot make.
        wants_list = any(
            expected[(name, row)].startswith(LIST_MARKER) for row in range(len(values))
        )
        if wants_list != is_list:
            print(
                f"{name}: the writer says {'a list' if wants_list else 'a scalar'} "
                f"and pyarrow read {field.type}"
            )
            return 1
        if is_list:
            lists += 1
            if not pa.types.is_string(field.type.value_type):
                print(f"{name}: list elements read as {field.type.value_type}, not string")
                return 1
            if field.type.value_field.nullable:
                print(f"{name}: list elements are nullable, and a validated token is not")
                return 1

        for row, actual in enumerate(values):
            want = expected[(name, row)]
            if is_list:
                if actual is None:
                    print(f"{name} row {row}: a list column must never be null")
                    return 1
                if any(e is None for e in actual):
                    print(f"{name} row {row}: a null element inside a list")
                    return 1
                got = LIST_MARKER + ",".join(actual)
            else:
                got = NULL if actual is None else str(actual)
            if got != want:
                print(f"{name} row {row}: file has {got!r}, writer meant {want!r}")
                return 1
            checked += 1

    print(
        f"pyarrow read {table.num_rows} rows, {len(table.column_names)} columns "
        f"({lists} of them lists), {checked} cells, all matching"
    )
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: crates/trailryx-projection/tests/oracle.rs
//! The check that the file really is Parquet.
//!
//! Everything else in this crate compares our writer against our own idea of
//! what it should produce, which is worth something and is not worth what this
//! is worth. Here the file goes to pyarrow, a reader written by other people,
//! and every cell comes back and is compared.
//!
//! It needs a Python with pyarrow, which not every machine has, so the test
//! reads a path from `TRAILRYX_PARQUET_ORACLE` and reports that it was skipped
//! when there is none. A skipped check that says so is honest; one that quietly
//! passes is the thing this whole project is against.
//!
//! ```text
//! python3 -m venv /tmp/oracle && /tmp/oracle/bin/pip install pyarrow
//! TRAILRYX_PARQUET_ORACLE=/tmp/oracle/bin/python cargo test -p trailryx-projection
//! ```

mod common;

use common::segment;
use std::process::Command;
use trailryx_projection::parquet::{Column, Values};
use trailryx_projection::{project, project_columns};

const NULL: &str = "\\0NULL";
/// A list cell is marked, so the oracle can insist the reader hands back a list
/// rather than a string that happens to render the same.
const LIST: &str = "\\0LIST";

#[test]
fn somebody_elses_reader_agrees_with_every_cell() {
    let Ok(python) = std::env::var("TRAILRYX_PARQUET_ORACLE") else {
        println!(
            "skipped: set TRAILRYX_PARQUET_ORACLE to a python with pyarrow to check the file \
             against a reader we did not write"
        );
        return;
    };

    let s = segment();
    let projection = project(&[&s]).unwrap();
    let columns = project_columns(&[&s]);

    let dir = std::env::temp_dir().join(format!("trailryx-oracle-{}", std::process::id()));
    std::fs::create_dir_all(&dir).unwrap();
    let parquet = dir.join("projection.parquet");
    let expected = dir.join("expected.tsv");
    std::fs::write(&parquet, projection.bytes()).unwrap();

    let mut table = String::new();
    for column in &columns {
        for row in 0..projection.rows() {
            let cell = match column.values.cell(row) {
                Some(rendered) if column.values.is_list_column() => format!("{LIST}{rendered}"),
                Some(rendered) => rendered,
                None => NULL.to_owned(),
            };
            table.push_str(&format!("{}\t{row}\t{cell}\n", column.name));
        }
    }
    std::fs::write(&expected, table).unwrap();

    let script = concat!(env!("CARGO_MANIFEST_DIR"), "/tests/oracle.py");
    let output = Command::new(python)
        .arg(script)
        .arg(&parquet)
        .arg(&expected)
        .output()
        .expect("the oracle should run");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let stderr = String::from_utf8_lossy(&output.stderr);
    assert!(
        output.status.success(),
        "pyarrow disagreed with the writer:\n{stdout}\n{stderr}"
    );
    println!("{stdout}");
    // The sibling test below always wiped its directory and this one never did, which
    // cost one stale directory while the path was a constant. With a process id in it,
    // it costs one on every run, so the two now behave the same way.
    let _ = std::fs::remove_dir_all(&dir);
}

/// The encoding hazard that lists exist to get wrong, put to an outside reader.
///
/// An empty list writes a level pair and **no value**. Get that wrong and every
/// later row's values shift by one, which produces a file that parses cleanly and
/// says something else. The pattern below is chosen so a shift of one cannot look
/// right anywhere: empty lists at the start, in the middle and at the end, next to
/// lists of one, two and three elements, with a scalar column beside them whose
/// values would visibly misalign.
#[test]
fn pyarrow_agrees_about_lists_with_empties_at_every_position() {
    let Ok(python) = std::env::var("TRAILRYX_PARQUET_ORACLE") else {
        println!("skipped: TRAILRYX_PARQUET_ORACLE is not set, so no outside reader saw the lists");
        return;
    };

    let lists: Vec<Vec<String>> = vec![
        vec![],
        vec!["a".into()],
        vec![],
        vec!["b".into(), "c".into()],
        vec!["d".into(), "e".into(), "f".into()],
        vec![],
        vec!["g".into()],
        vec![],
    ];
    let rows = lists.len();
    let columns = vec![
        Column::required(
            "n",
            Values::Int64((0..rows).map(|i| Some(i as i64 * 100)).collect()),
        ),
        Column::required("items", Values::StringList(lists.clone())),
        // A second list column after the first, because a mistake in the first
        // column's value count can also corrupt the offsets of the next one.
        Column::required(
            "again",
            Values::StringList(lists.iter().rev().cloned().collect()),
        ),
        Column::optional(
            "tail",
            Values::String((0..rows).map(|i| Some(format!("row-{i}"))).collect()),
        ),
    ];

    let bytes = trailryx_projection::parquet::write(&columns).expect("the file writes");
    let dir = std::env::temp_dir().join(format!("trailryx-oracle-lists-{}", std::process::id()));
    std::fs::create_dir_all(&dir).unwrap();
    let parquet = dir.join("lists.parquet");
    let expected = dir.join("expected.tsv");
    std::fs::write(&parquet, &bytes).unwrap();

    let mut table = String::new();
    for column in &columns {
        for row in 0..rows {
            let cell = match column.values.cell(row) {
                Some(rendered) if column.values.is_list_column() => format!("{LIST}{rendered}"),
                Some(rendered) => rendered,
                None => NULL.to_owned(),
            };
            table.push_str(&format!("{}\t{row}\t{cell}\n", column.name));
        }
    }
    std::fs::write(&expected, table).unwrap();

    let script = concat!(env!("CARGO_MANIFEST_DIR"), "/tests/oracle.py");
    let output = Command::new(python)
        .arg(script)
        .arg(&parquet)
        .arg(&expected)
        .output()
        .expect("the oracle should run");
    let stdout = String::from_utf8_lossy(&output.stdout);
    let stderr = String::from_utf8_lossy(&output.stderr);
    assert!(
        output.status.success(),
        "pyarrow disagreed about a list with empties:\n{stdout}\n{stderr}"
    );
    println!("{stdout}");
    let _ = std::fs::remove_dir_all(&dir);
}
